### PR TITLE
Copy py3k-sympy

### PR DESCRIPTION
**Pull request by asmeurer**
Automatically copy py3k-sympy when the `--reference` option is given.

This makes running the tests in Python 3 **way** faster if you have already run `./bin/use2to3` in the reference directory, because it will just have to run 2to3 on those files that the pull request changed (and those that changed since you last ran 2to3, so it's good to just run that often, to keep things up-to-date). 

You can explicitly disable this with the `--py3k-no-copy` option, which can also be set in the config file.  

For those running tests automatically on servers (@Krastanov and @certik), I recommend adding some kind of cron job to run use2to3 on your local SymPy periodically, to keep this fast.

### CODE REVIEW

**Comment by asmeurer** [comment]
This closes #58.

**Comment by asmeurer** [comment]
For some reason, I keep getting the following error:

``` pytb
Traceback (most recent call last):
  File "bin/use2to3", line 162, in <module>
    for line in fileinput.input(modified_files, inplace=1):
  File "/sw/lib/python3.2/fileinput.py", line 260, in __next__
    line = self.readline()
  File "/sw/lib/python3.2/fileinput.py", line 360, in readline
    return self.readline()
  File "/sw/lib/python3.2/fileinput.py", line 330, in readline
    self._file = open(self._backupfilename, self._mode)
IOError: [Errno 21] Is a directory: 'py3k-sympy/bin/__pycache__.bak'
```

This is fixed by manually removing that directory from my sympy repo.  But it seems to keep reappearing.  @vperic do you know anything about this?

**Comment by vperic** [comment]
Yeah, I noticed that too. I think the following is happening: the script traverses the tree and copies everything to the py3k-sympy/ dir, including all the **pycache** files. These then get passed to fileinput, which fixes the shebang lines.. but it obviously won't work with a directory. Still, files seem to be modified all right (verify with eg. grep -r "env python .) so I think it's harmless enough.

Solution would be to use git ls-files, but that pull is stalled because it doesn't work on Windows and I don't know how to fix it. Other solution would be to explicitly ignore **pycache** stuff somewhere earlier in the script.

**Comment by asmeurer** [comment]
Well, this should be as simple as just removing all the directories from `modified_files`, right?

**Comment by vperic** [comment]
Yeah, that'd work. I took a look, it should be enough to just add
"**pycache**.bak" to the skip_dirs list (**pycache**) is already
there.

On Thu, Dec 22, 2011 at 2:49 PM, Aaron Meurer
reply@reply.github.com
wrote:

> Well, this should be as simple as just removing all the directories from `modified_files`, right?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/sympy/sympy-bot/pull/82#issuecomment-3249143

## 

Vladimir Perić

File: README.rst [hunk 1]
```diff
@@ -35,13 +35,20 @@ Tips
...
+connection) by passing a ``--reference`` option to sympy-bot::
 
     ./sympy-bot --reference ~/repo/git/sympy review 268
 
 This gets passed to git, see ``git clone --help`` for more information. Then
 sympy-bot starts testing the branch immediately, even if you have a slower
-connections.
+connection.
+
+This has another advantage: if you run ``./use/2to3`` in the reference
```
Comment (vperic) [inline README.rst]: typo, it's bin/use2to3, not ./use/2to3

**Comment by asmeurer** [comment]
OK, this should **not** be merged until https://github.com/sympy/sympy/pull/726 is merged (that PR should also block on any release btw).  The reason is that you can get some false positive.  See for example https://github.com/sympy/sympy/pull/884#issuecomment-3271731.  Now, I don't actually know for sure if that pull request would have prevented that or not, but things like the `__pycache__` issue are fixed by it.

**Comment by certik** [comment]
I am ok with this patch. Feel free to merge if it works for you.

**Comment by asmeurer** [comment]
As I said before, this should not be merged, at least until sympy/sympy#726 is merged. Running `./bin/use2to3` on an already translated code-base is too unreliable. Even when that happens, I would like to test this a little more before releasing it.

**Comment by certik** [comment]
I am closing this pull request, so that it doesn't hang here if it is not ready. If you make it ready to merge, feel free to reopen it.

**Comment by asmeurer** [comment]
The work here is fine (except it does need to be merged/rebased).  It's just waiting on the SymPy pull request.  I'd rather not have it closed, or else it will be forgotten.

**Comment by certik** [comment]
Ok, in that case let's keep it open.

**Comment by flacjacket** [comment]
sympy/sympy#1349 (which replaced 726) is closed, so where does this stand?

**Comment by asmeurer** [comment]
I haven't had a chance to work out the merge issues. Feel free to pick it up if you want.

**Comment by asmeurer** [comment]
The code has changed so much since I did this that I have completely redone it.  See #125.  The option is now `--copy-py3k-sympy`, and it is disabled by default, but it can be added to the config file to enable it by default if you want.

**Pull request closed without merging**